aicsimageio/readers/mrc_reader.py

```diff
@@ -0,0 +1,149 @@
+        # Catch non-local file system
+        if not isinstance(self._fs, LocalFileSystem):
+            raise NotImplementedError(
+                f"mrc reader not yet implemented for non-local file system. "
+                f"Received URI: {self._path}, which points to {type(self._fs)}."
+            )
+
+        if not self._is_supported_image(self._fs, self._path):
+            raise exceptions.UnsupportedFileFormatError(
+                self.__class__.__name__, self._path
+            )
+
+    @property
+    def scenes(self) -> Tuple[str, ...]:
+        if self._scenes is None:
+            # need full mmap open here because `is_image_stack`
+            # needs access to data shape
+            with mrcfile.mmap(self._path, permissive=True) as mrc:
+                if mrc.is_image_stack():
+                    n_scenes = mrc.header.nz
+                else:
+                    n_scenes = mrc.header.nz / mrc.header.mz
+                self._scenes = tuple(
+                    generate_ome_image_id(i) for i in range(int(n_scenes))
+                )
+        return self._scenes
+
+    def _read_delayed(self) -> xr.DataArray:
+        return self._make_xarray(delayed=True)
+
+    def _read_immediate(self) -> xr.DataArray:
+        return self._make_xarray(delayed=False)
+
+    def _make_xarray(self, delayed: bool) -> xr.DataArray:
+        with mrcfile.mmap(self._path, permissive=True) as mrc:
+            if mrc.is_single_image():
+                scene = mrc.data
+                dims = [DimensionNames.SpatialY, DimensionNames.SpatialX]
+            elif mrc.is_image_stack():
+                scene = mrc.data[self._current_scene_index]
+                dims = [DimensionNames.SpatialY, DimensionNames.SpatialX]
+            elif mrc.is_volume():
+                scene = mrc.data
+                dims = [
+                    DimensionNames.SpatialZ,
+                    DimensionNames.SpatialY,
+                    DimensionNames.SpatialX,
+                ]
+            else:
+                scene = mrc.data[self._current_scene_index]
+                dims = [
+                    DimensionNames.SpatialZ,
+                    DimensionNames.SpatialY,
+                    DimensionNames.SpatialX,
+                ]
+
+            # build coordinates
+            coords = {}
+            for dim in dims:
+                mrc_dim = dim.lower()
+                start = mrc.header[f"n{mrc_dim}start"]
+                # only weird case, because header.mz is not 1 despite images having z=1
+                if mrc_dim == "z" and mrc.is_image_stack():
+                    end = 1
+                else:
+                    end = mrc.header[f"m{mrc_dim}"]
+                voxel_size = mrc.voxel_size[mrc_dim]
+                coords[dim] = self._generate_coord_array(start, end, voxel_size)
+
+            # convert before exiting the context manager or scene will be None
+            if delayed:
+                data = da.from_array(scene)
```

Comment:
> @tlambert03 do you have any suggestions on how to convert this code to use DaskArrayProxy to make the chunked reads pass file handling?

sure!  if this is the line where the dask array gets created, then I think you'd need to do a couple things:

make this line `data = DaskArrayProxy(da.from_array(scene), special_context_opener)`

the second argument there (`special_context_opener`) is an object that implements [this `CheckableContext` protocol here](https://github.com/AllenCellModeling/aicsimageio/blob/06ddd5e52d035ddcd711736959429adbcfca8336/aicsimageio/utils/dask_proxy.py#L17-L21), that is, it must have:
- a `closed` attribute that tells whether or not the underlying resource is currently closed
- an `__enter__` method that opens the resource
- an `__exit__` method that closes the resource

Have a look at `bioformats_reader.BioFile` for an example of how I did it there.  Essentially, the class itself was just a thing that manages the underlying file handle.  It implements the `CheckableContext` protocol, so I just pass it directly to `DaskArrayProxy(..., self)`

the main question (and I'd need to look closer to know) is whether you're going to need to get clever and add a layer on top of that base mrcfile reader or not.  Note that in `nd2`, `dv`, and `bioformats`, the base reader was ready for this approach, so the big question is whether or not `da.from_array(scene)` using the `mrcfile.mmap` is going to be sufficient.  @brisvag ... happy to zoom about this if you want

Reply:
I'm not sure I follow! What is the problem that needs solving exactly?

If I understand the last question correctly: `mrcfile.mmap` returns a `np.memmap`, which survives closing the resource (I suppose because it points directly to the address on the disk?).

```
import mrcfile
mrc = mrcfile.mmap('./resources/image.mrc')
data = mrc.data
mrc.close()
# data still works
```

Not sure how this fits in with the rest :P Yeah happy to have a quick call if I can't figure it out from these comments first!

Reply:
Tests are failing on Mac and Linux because there are file pointers left open after operations complete. DaskArrayProxy _should_ be able to help with closing them / handling them if I remember correctly.

Reply:
See test failures here: https://github.com/AllenCellModeling/aicsimageio/runs/4409557358?check_suite_focus=true

Reply:
welcome @brisvag to the challenging (but oh-so-admirable) constraints placed on aicsimageio reader :joy:

This turns out to be _really_ hard to do with lazy readers in a performant way.  Take the following thing:

```
with SomeReader('path') as reader:
    dsk = reader.to_dask()
```

`reader` should now be closed.  this means that the dask array `dsk` will _either_ need to know how to re-open the underlying resource (`SomeReader('path')`) whenever it needs to compute something, _or_ it needs to essentially close on a permanently opened file-handle/resource.  For good reasons, `aicsimageio` prohibits the latter.

A basic, easy(ish) to implement approach would be for that `to_dask()` function to return a `map_blocks`-like reader, that re-opens and closes the resource on _every_ chunk read.  but that can incur significant overhead if you have to open/close the file on every single chunk/plane in an nd stack.  `DaskArrayProxy` tries to solve this problem, by patching `Array.compute()` so that it opens the resource (_just once_) before any computations, and closes (if necessary) afterwards.  It's definitely experimental, but that's the story here.

-----

side-note for both of you:
I'm still trying to improve on this Proxy concept.  I'd like to embed it in the task graph itself rather than use this proxy object,  I asked a question on SO and got a good answer here: https://stackoverflow.com/questions/69967644/embedding-pre-post-compute-operations-in-a-dask-array-task-graph

Reply:
Ok, it's a bit clearer but still complicated. In practice, what do I have to implement? Something along these lines (thsi does not work, but also does not break more)?

```python
...
            if delayed:
                data = DaskArrayProxy(da.from_array(scene), CheckableMrc(mrc))
...
```

```python
class CheckableMrc:
    def __init__(self, mrc: mrcfile.mrcfile.MrcFile):
        self.mrc = mrc

    @property
    def closed(self):
        return self.mrc._iostream.closed  # this should work after __exit__ happened

    def __enter__(self):
        pass
        # somehow reopen the mrc file?

    def __exit__(self):
        self.mrc.close()  # this closes the iostream
```

Reply:
> to the challenging (but oh-so-admirable) constraints placed on aicsimageio reader

I _know_. Sometimes I hate this constraint and sometimes I appreciate it. I generally side on the appreciate it because it protects users from their pipelines breaking when trying to serialize the `AICSImage` / `Reader` objects when they are parallelizing their code. Open to discussion in _removing_ this constraint in v5 if we ever get there -- Off topic, anyway!

---

> I'm still trying to improve on this Proxy concept. I'd like to embed it in the task graph itself rather than use this proxy object, I asked a question on SO and got a good answer here: https://stackoverflow.com/questions/69967644/embedding-pre-post-compute-operations-in-a-dask-array-task-graph

That is a great answer.

Reply:
> Something along these lines (this does not work, but also does not break more stuff :P)?

yep! something along those lines :)